### PR TITLE
feat(kuma-cp) mesh insight

### PR DESCRIPTION
**Pull request by lobkovilya**
### Summary

Current PR adds `MeshInsight` entity that allows tracking number of mesh scoped resources.

#### API
- GET: /mesh-insights
```json
{
 "total": 2,
 "items": [
  {
   "type": "MeshInsight",
   "name": "default",
   "creationTime": "2020-11-11T07:07:41Z",
   "modificationTime": "2020-11-11T07:07:41Z",
   "lastSync": "2020-11-11T07:16:52.651679130Z",
   "dataplanes": {
    "total": 4,
    "online": 4
   },
   "policies": {
    "Config": {
     "total": 4
    },
    "TrafficPermission": {
     "total": 2
    }
   }
  },
  {
   "type": "MeshInsight",
   "name": "mymesh1",
   "creationTime": "2020-11-11T07:13:04Z",
   "modificationTime": "2020-11-11T07:13:04Z",
   "lastSync": "2020-11-11T07:16:52.662834520Z",
   "dataplanes": {},
   "policies": {
    "Config": {
     "total": 4
    },
    "Secret": {
     "total": 2
    },
    "TrafficPermission": {
     "total": 1
    }
   }
  }
 ],
 "next": null
}
```

- GET /mesh-insights/default
```json
{
 "type": "MeshInsight",
 "name": "default",
 "creationTime": "2020-11-11T07:07:41Z",
 "modificationTime": "2020-11-11T07:07:41Z",
 "lastSync": "2020-11-11T07:18:27.657468810Z",
 "dataplanes": {
  "total": 4,
  "online": 4
 },
 "policies": {
  "Config": {
   "total": 4
  },
  "TrafficPermission": {
   "total": 2
  }
 }
}
```

#### New command for `kumactl`:
```bash
$ kumactl inspect meshes
MESH      DATAPLANES   TRAFFIC PERMISSIONS   TRAFFIC ROUTES   CIRCUIT BREAKERS   HEALTH CHECKS   FAULT INJECTIONS   EXTERNAL SERVICES   TRAFFIC TRACES   TRAFFIC LOGS   PROXY TEMPLATES
default   4/4          2                     0                0                  0               0                  0                   0                0              0
mymesh1   0/0          1                     0                0                  0               0                  0                   0                0              0
```
### Full changelog

* New event-based approach. We don't recalculate MeshInsight often. We do that when an event happens, but not more often than `minResyncTimeout`(using rateLimiter) and every `maxResyncTimeout` to keep it synced in case of failures. 

### Issues resolved

Fix #XXX

### Documentation

- [ ] Link to the website [documentation PR](https://github.com/kumahq/kuma-website/pull/XXX)

### CODE REVIEW

**Comment by subnetmarco** [comment]
In the API this is `/mesh-insights`, but the other insights I believe are `/dataplane+insights`. Shouldn't we have some consistency here?

File: pkg/events/eventbus.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+type EventBus struct {
+	mtx         sync.RWMutex
+	subscribers []chan Event
+}
+
+func (b *EventBus) New() Reader {
+	b.mtx.Lock()
+	defer b.mtx.Unlock()
+
+	events := make(chan Event, 10)
```
Comment (jakubdyszkiewicz) [inline pkg/events/eventbus.go:24]: I'm just slightly concerned about this `10`. What will happen if one consumer of events won't process the events quickly enough? I think the sender will be stuck at `Send` operation right? What then happens in Postgres / K8S? Will we lose events or not?

File: pkg/events/interfaces.go [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+	Create Op = iota
+	Update
+	Delete
+)
+
+type Event struct {
+	Operation Op
+	Type      model.ResourceType
+	Key       model.ResourceKey
+}
```
[the remark this answers is not in the document]
Comment (jakubdyszkiewicz) [reply pkg/events/interfaces.go:19]: Also, please note as a comment that `ResourceChangedEvent` is emitted on every instance of the control plane

**Review by jakubdyszkiewicz — CHANGES_REQUESTED** [review]
nice, I like the event-based approach, I think it has a lot of potential in the future.

File: pkg/insights/resyncer.go [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+		}
+	}(stop)
+
+	limiter := rate.NewLimiter(rate.Every(p.minResyncTimeout), 50)
+	for {
+		event, err := p.events.Recv(stop)
+		if err != nil {
+			return err
+		}
+		if !meshScoped(event.Type) || event.Type == core_mesh.DataplaneType {
```
[the remark this answers is not in the document]
Comment (lobkovilya) [reply pkg/insights/resyncer.go:82]: They are 1:1 to DataplaneInsights, we need DataplaneInsight to check online/offline status, so Dataplane itself is kind of useless for counting

File: pkg/insights/resyncer.go [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+				}
+			case <-stop:
+				log.Info("stop")
+				return
+			}
+		}
+	}(stop)
+
+	limiter := rate.NewLimiter(rate.Every(p.minResyncTimeout), 50)
+	for {
```
[the remark this answers is not in the document]
Comment (lobkovilya) [reply pkg/insights/resyncer.go:77]: yeah, I pass `stop` channel into `Recv` to gracefully finish when CP is exiting

File: api/mesh/v1alpha1/mesh_insight.proto [hunk 5]
```diff
@@ -0,0 +1,19 @@
+syntax = "proto3";
+
+package kuma.mesh.v1alpha1;
+
+import "google/protobuf/timestamp.proto";
+
+message MeshInsight {
```
[the remark this answers is not in the document]
Comment (lobkovilya) [reply api/mesh/v1alpha1/mesh_insight.proto:8]: Added docs

File: app/kumactl/cmd/inspect/inspect_meshes.go [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+
+func newInspectMeshesCmd(ctx *inspectContext) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "meshes",
+		Short: "Inspect Meshes",
+		Long:  `Inspect Meshes.`,
+		RunE: func(cmd *cobra.Command, _ []string) error {
+			client, err := ctx.CurrentResourceStore()
+			if err != nil {
+				return errors.Wrap(err, "failed to create a zone client")
```
[the remark this answers is not in the document]
Comment (lobkovilya) [reply app/kumactl/cmd/inspect/inspect_meshes.go]: Fixed, thanks

File: pkg/config/app/kuma-cp/config.go [hunk 7]
```diff
@@ -81,6 +82,21 @@ func (d *ZoneMetrics) Validate() error {
 	return nil
 }
 
+type MeshMetrics struct {
+	MinResyncTimeout time.Duration `yaml:"minResyncTimeout" envconfig:"kuma_metrics_mesh_min_resync_timeout"`
+	MaxResyncTimeout time.Duration `yaml:"maxResyncTimeout" envconfig:"kuma_metrics_mesh_max_resync_timeout"`
```
[the remark this answers is not in the document]
Comment (lobkovilya) [reply pkg/config/app/kuma-cp/config.go:89]: Commented, thanks

File: pkg/plugins/resources/postgres/events/listener.go [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+			case "UPDATE":
+				op = events.Update
+			case "DELETE":
+				op = events.Delete
+			default:
+				log.Error(errors.New("unknown Action"), "failed to parse action", "action", op)
+				continue
+			}
+			k.out.Send(op, model.ResourceType(obj.Data.Type), model.ResourceKey{Mesh: obj.Data.Mesh, Name: obj.Data.Name})
+		case <-time.After(90 * time.Second):
```
[the remark this answers is not in the document]
Comment (lobkovilya) [reply pkg/plugins/resources/postgres/events/listener.go]: No need if we log reconnect event, deleted this

File: pkg/config/app/kuma-cp/kuma-cp.defaults.yaml [hunk 9]
```diff
@@ -247,6 +247,11 @@ metrics:
     enabled: true # ENV: KUMA_METRICS_ZONE_ENABLED
     # How many latest subscriptions will be stored in ZoneInsights object, if equals 0 then unlimited
     subscriptionLimit: 10 # ENV: KUMA_METRICS_ZONE_SUBSCRIPTION_LIMIT
+  mesh:
+    # Min time that should pass between Mesh resync
```
[the remark this answers is not in the document]
Comment (lobkovilya) [reply pkg/config/app/kuma-cp/kuma-cp.defaults.yaml]: Fixed, thanks!

File: pkg/plugins/resources/postgres/events/listener.go [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+	"github.com/kumahq/kuma/pkg/core"
+	"github.com/kumahq/kuma/pkg/core/resources/model"
+	"github.com/kumahq/kuma/pkg/core/runtime/component"
+	"github.com/kumahq/kuma/pkg/events"
+	common_postgres "github.com/kumahq/kuma/pkg/plugins/common/postgres"
+)
+
+var log = core.Log.WithName("postgres-event-listener")
+
+type listener struct {
```
[the remark this answers is not in the document]
Comment (lobkovilya) [reply pkg/plugins/resources/postgres/events/listener.go:18]: Yes, it reconnects itself, sends `nil` into the channel, and then start sending events again

File: pkg/plugins/resources/k8s/events/listener.go [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+	if err != nil {
+		return nil, err
+	}
+	listGVK := gvk.GroupVersion().WithKind(gvk.Kind + "List")
+	listObj, err := k.mgr.GetScheme().New(listGVK)
+	if err != nil {
+		return nil, err
+	}
+	paramCodec := runtime.NewParameterCodec(k.mgr.GetScheme())
+	ctx := context.TODO()
```
[the remark this answers is not in the document]
Comment (lobkovilya) [reply pkg/plugins/resources/k8s/events/listener.go]: Fixed, thanks

File: pkg/plugins/resources/k8s/events/listener.go [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+		go func() {
+			informer.Run(stop)
+		}()
+	}
+	return nil
+}
+
+func resourceKey(obj model.KubernetesObject) core_model.ResourceKey {
+	var name string
+	if obj.GetNamespace() == "" {
```
[the remark this answers is not in the document]
Comment (lobkovilya) [reply pkg/plugins/resources/k8s/events/listener.go]: Fixed, thanks

File: pkg/plugins/resources/k8s/events/listener.go [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+}
+
+func (k *listener) NeedLeaderElection() bool {
+	return false
+}
+
+func (k *listener) addTypeInformationToObject(obj runtime.Object) error {
+	gvks, _, err := k.mgr.GetScheme().ObjectKinds(obj)
+	if err != nil {
+		return fmt.Errorf("missing apiVersion or kind and cannot assign it; %w", err)
```
[the remark this answers is not in the document]
Comment (lobkovilya) [reply pkg/plugins/resources/k8s/events/listener.go]: Fixed, thanks

File: pkg/plugins/resources/k8s/events/listener.go [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+		coreObj, err := core_registry.Global().NewObject(t)
+		if err != nil {
+			return err
+		}
+		obj, err := k8s_registry.Global().NewObject(coreObj.GetSpec())
+		if err != nil {
+			return err
+		}
+		informer := cache.NewSharedInformer(lw, obj, 0)
+		informer.AddEventHandler(k)
```
[the remark this answers is not in the document]
Comment (lobkovilya) [reply pkg/plugins/resources/k8s/events/listener.go:64]: Added log, thanks

File: pkg/plugins/common/postgres/listener.go [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+	if err != nil {
+		return nil, err
+	}
+	reportProblem := func(ev pq.ListenerEventType, err error) {
+		if err != nil {
+			log.Error(err, "error occurred")
+		}
+	}
+	listener := pq.NewListener(connStr, 10*time.Second, time.Minute, reportProblem)
+	if err := listener.Listen("events"); err != nil {
```
[the remark this answers is not in the document]
Comment (lobkovilya) [reply pkg/plugins/common/postgres/listener.go]: Yep, we can, renamed to `resource_events`

File: pkg/plugins/common/postgres/listener.go [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+
+	config "github.com/kumahq/kuma/pkg/config/plugins/resources/postgres"
+)
+
+func NewListener(cfg config.PostgresStoreConfig, log logr.Logger) (*pq.Listener, error) {
+	connStr, err := connectionString(cfg)
+	if err != nil {
+		return nil, err
+	}
+	reportProblem := func(ev pq.ListenerEventType, err error) {
```
[the remark this answers is not in the document]
Comment (lobkovilya) [reply pkg/plugins/common/postgres/listener.go:17]: Now we are logging them all

File: pkg/plugins/common/postgres/listener.go [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+		if err != nil {
+			log.Error(err, "error occurred")
+		}
+	}
+	listener := pq.NewListener(connStr, 10*time.Second, time.Minute, reportProblem)
```
[the remark this answers is not in the document]
Comment (lobkovilya) [reply pkg/plugins/common/postgres/listener.go]: Yes, totally agree, it also calms my sleep :) Fixed

File: pkg/insights/resyncer.go [hunk 18]
```diff
@@ -0,0 +1,189 @@
...
+		event, err := p.events.Recv(stop)
+		if err != nil {
+			return err
```
[the remark this answers is not in the document]
Comment (lobkovilya) [reply pkg/insights/resyncer.go:90]: Technically we can't read from `event` again if an error returned. An error means the channel is closed. I moved the event factory right here and wrap this component with resilientComponent

File: pkg/insights/resyncer.go [hunk 19]
```diff
@@ -0,0 +1,189 @@
...
+
+type resyncer struct {
+	rm               manager.ResourceManager
+	events           events.Reader
+	minResyncTimeout time.Duration
+	maxResyncTimeout time.Duration
+	tick             func(d time.Duration) <-chan time.Time
+}
+
+func NewResyncer(config *Config) component.Component {
```
[the remark this answers is not in the document]
Comment (lobkovilya) [reply pkg/insights/resyncer.go:52]: Added a docs

File: pkg/insights/resyncer.go [hunk 20]
```diff
@@ -0,0 +1,189 @@
+package insights
```
[the remark this answers is not in the document]
Comment (lobkovilya) [reply pkg/insights/resyncer.go:1]: I had an idea to use the same component for other `insights` as well. What do you think?

File: app/kuma-cp/cmd/run.go [hunk 21]
```diff
@@ -6,6 +6,8 @@ import (
 
 	"github.com/spf13/cobra"
 
+	"github.com/kumahq/kuma/pkg/insights"
```
[the remark this answers is not in the document]
Comment (lobkovilya) [reply app/kuma-cp/cmd/run.go]: Fixed, thanks

File: app/kuma-cp/cmd/run.go [hunk 22]
```diff
@@ -146,6 +148,11 @@ func newRunCmdWithOpts(opts runCmdOpts) *cobra.Command {
 				return err
 			}
 
+			if err := insights.Setup(rt); err != nil {
+				runLog.Error(err, "unable to set up Insights resyncer")
+				return err
+			}
```
[the remark this answers is not in the document]
Comment (lobkovilya) [reply app/kuma-cp/cmd/run.go]: Yes, moved it into Standalone and Global section

File: pkg/insights/resyncer.go [hunk 23]
```diff
@@ -0,0 +1,189 @@
...
+			}
+		}
+	}
+
+	if err := manager.Upsert(p.rm, model.ResourceKey{Mesh: model.NoMesh, Name: mesh}, &core_mesh.MeshInsightResource{}, func(resource model.Resource) {
+		insight.LastSync = proto.MustTimestampProto(core.Now())
+		_ = resource.SetSpec(insight)
+	}); err != nil {
+		return err
+	}
```
[the remark this answers is not in the document]
Comment (lobkovilya) [reply pkg/insights/resyncer.go:183]: Added ownership, thanks!

File: pkg/events/interfaces.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (lobkovilya) [reply pkg/events/interfaces.go:21]: Yes, that's a good idea! I did like you proposed, thanks!

File: pkg/events/interfaces.go [hunk 24]
```diff
@@ -0,0 +1,31 @@
...
+
+type Reader interface {
+	Recv(stop <-chan struct{}) (Event, error)
+}
+
+type Writer interface {
+	Send(Op, model.ResourceType, model.ResourceKey)
+}
```
[the remark this answers is not in the document]
Comment (lobkovilya) [reply pkg/events/interfaces.go:29]: I used `Reader/Writer` interfaces everywhere where it's possible to split the component's users into one who passively reads and the one who actively writes something. Any way you use it through the package: `events.Writer`, so it's pretty clear. 
But `Listener/Emitter` probably suits better here, renamed.

File: pkg/events/eventbus.go (hunk 1 above)
Comment (lobkovilya) [reply pkg/events/eventbus.go:22]: I looked at the implementation of Kubernetes/Postgress components that send events to EventBus. They have their own queues/channels and from what I saw they don't drop events. And right now we have only one consumer - `resyncer`. It uses `rateLimiter` and will drop events itself if will be too loaded.

File: pkg/insights/resyncer.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (jakubdyszkiewicz) [reply pkg/insights/resyncer.go]: I don't think they are 1:1. On Universal, we still support the flow when we create a dataplane before connecting kuma dp to cp. This way we don't have insights yet

File: pkg/insights/resyncer.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (jakubdyszkiewicz) [reply pkg/insights/resyncer.go:1]: Ok, let's bring it up again when we will have another insights

File: pkg/insights/resyncer.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (lobkovilya) [reply pkg/insights/resyncer.go]: Yes, you are right! Fixed

**Review by jakubdyszkiewicz — APPROVED** [review]

**Pull request merged**